components/sup/src/manager/commands.rs

```diff
@@ -195,22 +182,27 @@ pub fn service_file_put(
         filename,
         service_group,
     );
-    let mut client = match butterfly::client::Client::new(
-        mgr.cfg.gossip_listen.local_addr(),
-        mgr.cfg.ring_key.clone(),
-    ) {
-        Ok(client) => client,
-        Err(err) => {
-            outputln!("Failed to connect to own gossip server, {}", err);
-            return Err(net::err(ErrCode::Internal, err.to_string()));
-        }
-    };
-    match client.send_service_file(service_group, filename, version, &content, is_encrypted) {
-        Ok(()) => {
-            req.reply_complete(net::ok());
-            return Ok(());
+    create_butterfly_client(mgr)?
+        .send_service_file(service_group, filename, version, &content, is_encrypted)
+        .map(|_| req.reply_complete(net::ok()))
+        .map_err(|e| net::err(ErrCode::Internal, e.to_string()))
```

Comment:
I like the way you refactored this to reduce the code duplication! I think you could go even further, since 3 of these 4 lines are repeated 3 times. In order to put the emphasis on what's *different*, you could further factor out the rest of the error handling. I poked at it a bit and was able to get this to compile:
```rust
    with_butterfly_client(mgr, req, |mut client| {
        client.send_service_file(
            service_group.clone(),
            filename.clone(),
            version,
            &content,
            is_encrypted,
        )
    })
```
where `with_butterfly_client` is like:
```rust
fn with_butterfly_client(
    mgr: &ManagerState,
    req: &mut CtlRequest,
    f: impl Fn(butterfly::client::Client) -> butterfly::error::Result<()>,
) -> NetResult<()> {
    … // basically what you had already inside
```
With a bit more work, I imagine you could get rid of the clones as well, though for such small pieces of data attached to user-initiated commands, it may not be worthwhile.

However, there's an alternative to consider here: is this the right level to put the conditional handling of `mgr.cfg.gossip_listen`? Conceptually, a single check for whether we have a gossip server running at the point commands are dispatched seems to reduce the complexity, though if you looked at https://github.com/habitat-sh/habitat/blob/master/components/sup/src/ctl_gateway/server.rs#L303, I can certainly understand why you didn't go that route: it's far from obvious to see how to get at `ManagerState` at that level. Given the state of that code (which @christophermaier is in the midst of refactoring), I think your current approach is pretty reasonable, but I'd like to see an issue filed to revisit this once the control gateway code has been simplified because I think that's the more appropriate level to handle this.